posts/python-decorators.md

```diff
@@ -0,0 +1,180 @@
+def hello():
+    print('hello')
+
+>>> hello()
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+TypeError: 'int' object is not callable
+'int' object is not callable
+
+>>> hello
+5
+```
+
+
+# Writing your own decorator
+As mentioned, a decorator is simply a function that is passed a function, and returns an object. So, to start writing a decorator, we just need to define a function. In most cases, we want the object our decorator returns, to actually mimic the function we decorated. This means that the object the decorator returns, needs to be a function itself. For example, let's say we wanted to simply print every time the function was called, we could write a function that prints that information, then calls the function. But that function would need to be returned by the decorator. This usually leads to function nesting such as:
+
+```python
+def mydecorator(f):  # f is the function passed to us from python
+    def log_f_as_called():
+        print(f'{f} was called.')
+        f()
+    return log_f_as_called
+```
+
+As you can see, we have a nested function definition, and the decorator function is returning that function. This way, the function `hello` can still be called like a standard function, the caller does not need to know that it is decorated. We can now define `hello` as the following:
+
+```python
+@mydecorator
+def hello():
+    print('hello')
+```
+
+We get the following output:
+```python
+>>> hello()
+called <function hello at 0x7f27738d7510>
+hello
+```
+
+You can also decorate a function many times, to do many things if you wanted, in which case, its just a chain effect. Essentially, the top decorator is passed the object from the former, and so on and so forth. If we have the following code:
+
+```python
+@a
+@b
+@c
+def hello():
+    print('hello')
+```
+The interpreter is essentially doing `hello = a(b(c(hello)))` and all of the decorators would wrap each other. You can test this out yourself by using our existing decorator, and using it twice.
+
+```python
+@mydecorator
+@mydecorator
+def hello():
+    print('hello')
+
+>>> hello()
+called <function mydec.<locals>.a at 0x7f277383d378>
+called <function hello at 0x7f2772f78ae8>
+hello
+```
+
+You will notice that the first decorator, wrapped the second, and logged it separately. 
+
+One interesting thing here, is that once `hello` was wrapped once, it no longer appeared as hello to the interpreter. This is fine for our trivial example, but can often break tests and things that might be trying to introspect the function attributes. If the idea is for a decorator to act like the function it decorates, it needs to also mimic that function. Luckily, there is a python standard library decorator called `wraps` for that in functools module. It tells the function to appear as the function it wraps. 
+
+```python
+import functools
+def mydecorator(f): 
+    @functools.wraps(f)  # we tell wraps that the function we are wrapping is f
```

Comment:
I think it would be helpful to describe this as "functools.wraps is a _decorator_"